rfcs/convergence/slot-null-rendering.md

```diff
@@ -0,0 +1,232 @@
+
+1. `getSlots` should stop verifying null rendering by the existence of the property `children`
+2. Since some slots have default props, verification must be done on shorthand declaration to ensure that those slots will be null rendered (or not) even presenting default props
+
+## Background
+
+`getSlots` is a method that iterates over `ObjectShorthandProps` (_`shorthands`_) declarations and converts them to something that can be rendered by React.
+
+In the case where one _shorthand_ declares a native element (e.g: `"div"`, `"input"`) a verification is made to decide if this declaration will be rendered into something or not (null rendering). This verification will only allow native elements to be rendered if this element presents `children` as property.
+
+```tsx
+// react-utilities/src/compose/getSlots.ts
+// ....
+if (typeof slot === 'string' && children === undefined) {
+  slot = nullRender;
+}
+```
+
+## Problem statement
+
+The verification of only allowing native elements to be rendered if this element present `children` as property breaks in some edge cases
+
+### [Empty Elements]
+
+The `Input` component is a good case example of that as [reported](https://github.com/microsoft/fluentui/pull/18642#issuecomment-871677964) by @ecraig12345, since the native element `input` is an [Empty Element], it cannot take any children, and still need to be rendered.
+
+Right now for input case this is the solution to avoid `children = undefined` problem:
+
+```tsx
+export function Input(props) {
+  const state = {
+    components: {
+      input: 'input'
+    },
+    input: resolveShorthand(props.input, {
+      children: React.Fragment // 🚨 getSlots requires children
+    })
+  };
+  const { slots, slotProps } = getSlots(state, ['input']);
+  delete slotProps.input.children; // 🚨 input can't have children
+
+  return (
+    <slots.root {...slotProps.root}>
+      <slots.input {...slotProps.input} />
+    </slots.root>
+  );
+}
+```
+[Live Example](https://stackblitz.com/edit/react-ts-wwguyp?file=Input.tsx)
+
+
+### Slot as a parent for other slots
+
+This edge case has been seen in different converged components ([`CompoundButton`](https://github.com/microsoft/fluentui/blob/master/packages/react-button/src/components/CompoundButton/renderCompoundButton.tsx#L18), [`AccordionHeader`](https://github.com/microsoft/fluentui/blob/master/packages/react-accordion/src/components/AccordionHeader/renderAccordionHeader.tsx#L15))
+
+The problem is that using a slot as parent for other slots means overriding `children` props, so `children` is not declared in the moment `resolveShorthand` is invoked but in the moment that the slot will be rendered.
+
+
+```tsx
+export function Component(props) {
+  const state = {
+    components: {
+      button: 'button',
+      icon: 'i'
+    },
+    button: resolveShorthand(props.button, {
+      children: React.Fragment // 🚨 getSlots requires children
```

Comment:
Makes more sense to have this as some more meaningful chldren than a fragment in this example IMO

Reply:
It could be anything but undefined, I'll update it 👍🏼